### PR TITLE
[lldb][FrameRecognizer] Display the first non-std frame on verbose_trap

### PR DESCRIPTION
**Pull request by Michael137**
This attempts to improve user-experience when LLDB stops on a verbose_trap. Currently if a `__builtin_verbose_trap` triggers, we display the first frame above the call to the verbose_trap. So in the newly added test case, we would've previously stopped here:
```
(lldb) run
Process 28095 launched: '/Users/michaelbuch/a.out' (arm64)
Process 28095 stopped
* thread #1, queue = 'com.apple.main-thread', stop reason = Bounds error: out-of-bounds access
    frame #1: 0x0000000100003f5c a.out`std::__1::vector<int>::operator[](this=0x000000016fdfebef size=0, (null)=10) at verbose_trap.cpp:6:9
   3    template <typename T>
   4    struct vector {
   5        void operator[](unsigned) {
-> 6            __builtin_verbose_trap("Bounds error", "out-of-bounds access");
   7        }
   8    };
```

After this patch, we would stop in the first non-`std` frame:
```
(lldb) run
Process 27843 launched: '/Users/michaelbuch/a.out' (arm64)
Process 27843 stopped
* thread #1, queue = 'com.apple.main-thread', stop reason = Bounds error: out-of-bounds access
    frame #2: 0x0000000100003f44 a.out`g() at verbose_trap.cpp:14:5
   11  
   12   void g() {
   13       std::vector<int> v;
-> 14       v[10];
   15   }
   16  
```

rdar://134490328

### CODE REVIEW

**Review by ldionne — COMMENTED** [review]
I'm not qualified to review the implementation, but I like the result!

**Comment by egorzhdan** [comment]
This is a great UX improvement, thanks!

File: lldb/source/Target/VerboseTrapFrameRecognizer.cpp [hunk 1]
```diff
@@ -16,6 +16,31 @@ using namespace llvm;
 using namespace lldb;
 using namespace lldb_private;
 
+/// The 0th frame is the artificial inline frame generated to store
+/// the verbose_trap message. So, starting with the current parent frame,
+/// find the first frame that's not inside of the STL.
+static StackFrameSP FindMostRelevantFrame(Thread &selected_thread) {
+  StackFrameSP most_relevant_frame_sp = selected_thread.GetStackFrameAtIndex(1);
+  while (most_relevant_frame_sp) {
```
Comment (adrian-prantl) [inline lldb/source/Target/VerboseTrapFrameRecognizer.cpp:24]: can you add an upper bound for the loop? I'm afraid that this recognizer might be triggered when looking at the backtrace of an infinite recursion, for example, and then LLDB just hangs.

File: lldb/test/Shell/Recognizer/verbose_trap-in-stl-nested.test [hunk 2]
```diff
@@ -0,0 +1,13 @@
+# Tests that we show the first non-STL frame when
+# a verbose_trap triggers from within the STL.
+
+# UNSUPPORTED: system-windows
+#
+# RUN: %clang_host -g -O0 %S/Inputs/verbose_trap-in-stl-nested.cpp -o %t.out
+# RUN: %lldb -b -s %s %t.out | FileCheck %s --check-prefixes=CHECK
```
Comment (adrian-prantl) [inline lldb/test/Shell/Recognizer/verbose_trap-in-stl-nested.test:7]: Is there a good reason why this isn't an API test?

Comment (Michael137) [reply lldb/test/Shell/Recognizer/verbose_trap-in-stl-nested.test:7]: I find it much easier to understand what the test is trying to check when it's done as a `Shell` test. Especially for when we just want to inspect the format of the frame. (FWIW, all the other frame format tests are also `Shell` tests).

From the [Testing FAQ](https://lldb.llvm.org/resources/test.html):
> API tests: Integration tests that interact with the debugger through the SB API. These are written in Python and use LLDB’s dotest.py testing framework on top of Python’s [unittest](https://docs.python.org/3/library/unittest.html).

And later:
> A good rule of thumb is to prefer shell tests when what is being tested is relatively simple. Expressivity is limited compared to the API tests, which means that you have to have a well-defined test scenario that you can easily match with FileCheck.

IMO these tests fit into the latter category.

But I'm happy to change these to API tests if that is more appropriate.

**Review by adrian-prantl — APPROVED** [review]

**Pull request merged**